### PR TITLE
[MOSIP-34112] Release 1.2.0.x code marge for Bio-dedupes updated flow

### PR DESCRIPTION
**Pull request by Khuddusshariff0022**

### CODE REVIEW

**Review by ashok-ksharma — APPROVED** [review]
Approved

**Review by ashok-ksharma — APPROVED** [review]

File: registration-processor/core-processor/registration-processor-bio-dedupe-stage/src/main/java/io/mosip/registration/processor/biodedupe/stage/BioDedupeProcessor.java [hunk 1]
```diff
@@ -435,24 +427,81 @@ private void postAbisIdentification(InternalRegistrationStatusDto registrationSt
...
+						matchedRegIds);
+			}else {
+			    sendToManualAdjudicationStage(registrationStatusDto, object, registrationType, moduleName, matchedRegIds);
+			    regProcLogger.info(LoggerFileConstant.SESSIONID.toString(), LoggerFileConstant.REGISTRATIONID.toString(),
 					registrationStatusDto.getRegistrationId(), BioDedupeConstants.ABIS_RESPONSE_NOT_NULL);
+			}
+		}
+	}
 
+	private void updatePostAbisIdentification(InternalRegistrationStatusDto registrationStatusDto, MessageDTO object,
```
Comment (vishwa-vyom) [inline registration-processor/core-processor/registration-processor-bio-dedupe-stage/src/main/java/io/mosip/registration/processor/biodedupe/stage/BioDedupeProcessor.java]: Lets rename this to handlePostAbisIdentificationForUpdate

File: registration-processor/core-processor/registration-processor-bio-dedupe-stage/src/test/java/io/mosip/registration/processor/biodedupe/stage/BioDedupeProcessorTest.java [hunk 2]
```diff
@@ -85,604 +91,833 @@
...
+    @Test
+    public void noMatchedRidFound() throws ApisResourceAccessException, IOException,
+            PacketDecryptionFailureException, io.mosip.kernel.core.exception.IOException, JsonProcessingException, PacketManagerException {
+        registrationStatusDto.setRegistrationId("27847657360002520190320095011");
+        registrationStatusDto.setRegistrationType("UPDATE");
+        registrationStatusDto.setStatusCode("PROCESSING");
+        Mockito.when(registrationStatusService.getRegistrationStatus(any(),any(),any(), any())).thenReturn(registrationStatusDto);
+        Mockito.when(abisHandlerUtil.getPacketStatus(any())).thenReturn(AbisConstant.POST_ABIS_IDENTIFICATION);
+        Set<String> matchedRidList = new HashSet<>();
+        Mockito.when(abisHandlerUtil.getUniqueRegIds(any(), any(), anyInt(), any(), any())).thenReturn(matchedRidList);
```
Comment (vishwa-vyom) [inline registration-processor/core-processor/registration-processor-bio-dedupe-stage/src/test/java/io/mosip/registration/processor/biodedupe/stage/BioDedupeProcessorTest.java:719]: Can we have one more test case with matchedRidList as null ?

File: registration-processor/core-processor/registration-processor-bio-dedupe-stage/src/test/java/io/mosip/registration/processor/biodedupe/stage/BioDedupeProcessorTest.java [hunk 3]
```diff
@@ -85,604 +91,833 @@
...
+        Set<String> matchedRidList = new HashSet<>();
+        matchedRidList.add("27847657360002520190320095011");
+        Mockito.when(abisHandlerUtil.getUniqueRegIds(any(), any(), anyInt(), any(), any())).thenReturn(matchedRidList);
+        Mockito.when(idRepoService.getUinByRid(anyString(),any())).thenReturn("9403107397");
+        ReflectionTestUtils.setField(bioDedupeProcessor, "biometricsUpdateSingleMatchAutoReject", false);
+        ArgumentCaptor<InternalRegistrationStatusDto> argumentCaptor = ArgumentCaptor.forClass(InternalRegistrationStatusDto.class);
+        MessageDTO messageDto = bioDedupeProcessor.process(dto, stageName);
+        Mockito.verify(registrationStatusService).updateRegistrationStatus(argumentCaptor.capture(),anyString(),anyString());
+        assertTrue(messageDto.getIsValid());
+        assertFalse(messageDto.getInternalError());
```
Comment (vishwa-vyom) [inline registration-processor/core-processor/registration-processor-bio-dedupe-stage/src/test/java/io/mosip/registration/processor/biodedupe/stage/BioDedupeProcessorTest.java:782]: Can we also check the assert on setMessageBusAddress to make sure the manual adjudication is given ?

Comment (vishwa-vyom) [reply registration-processor/core-processor/registration-processor-bio-dedupe-stage/src/test/java/io/mosip/registration/processor/biodedupe/stage/BioDedupeProcessorTest.java:782]: lets follow this for all case where manual adjudication needs to be done

File: registration-processor/core-processor/registration-processor-bio-dedupe-stage/src/test/java/io/mosip/registration/processor/biodedupe/stage/BioDedupeProcessorTest.java [hunk 4]
```diff
@@ -85,604 +91,833 @@
...
+        Mockito.when(abisHandlerUtil.getUniqueRegIds(any(), any(), anyInt(), any(), any())).thenReturn(matchedRidList);
+        Mockito.when(idRepoService.getUinByRid(anyString(),any()))
+                .thenReturn("9403107397")
+                .thenReturn("9403107399");
+        ArgumentCaptor<InternalRegistrationStatusDto> argumentCaptor = ArgumentCaptor.forClass(InternalRegistrationStatusDto.class);
+        MessageDTO messageDto = bioDedupeProcessor.process(dto, stageName);
+        Mockito.verify(registrationStatusService).updateRegistrationStatus(argumentCaptor.capture(),anyString(),anyString());
+        assertTrue(messageDto.getIsValid());
+        assertFalse(messageDto.getInternalError());
+        Assert.assertEquals(argumentCaptor.getValue().getStatusCode(), RegistrationStatusCode.FAILED.name());
```
Comment (vishwa-vyom) [inline registration-processor/core-processor/registration-processor-bio-dedupe-stage/src/test/java/io/mosip/registration/processor/biodedupe/stage/BioDedupeProcessorTest.java:820]: In this and most of the below negative cases biometricsUpdateSingleMatchAutoReject should be set to true right ?

**Review by vishwa-vyom — CHANGES_REQUESTED** [review]

File: registration-processor/core-processor/registration-processor-bio-dedupe-stage/src/main/java/io/mosip/registration/processor/biodedupe/stage/BioDedupeProcessor.java (hunk 1 above)
Comment (Khuddusshariff0022) [reply registration-processor/core-processor/registration-processor-bio-dedupe-stage/src/main/java/io/mosip/registration/processor/biodedupe/stage/BioDedupeProcessor.java]: Done.

File: registration-processor/core-processor/registration-processor-bio-dedupe-stage/src/test/java/io/mosip/registration/processor/biodedupe/stage/BioDedupeProcessorTest.java (hunk 2 above)
Comment (Khuddusshariff0022) [reply registration-processor/core-processor/registration-processor-bio-dedupe-stage/src/test/java/io/mosip/registration/processor/biodedupe/stage/BioDedupeProcessorTest.java:719]: This is not possible as per Code. Because the method will not return null instead it will return empty set. if we do then we will get "Unknown exception occurred null" Exception error.

File: registration-processor/core-processor/registration-processor-bio-dedupe-stage/src/test/java/io/mosip/registration/processor/biodedupe/stage/BioDedupeProcessorTest.java (hunk 3 above)
Comment (Khuddusshariff0022) [reply registration-processor/core-processor/registration-processor-bio-dedupe-stage/src/test/java/io/mosip/registration/processor/biodedupe/stage/BioDedupeProcessorTest.java:782]: Done.

File: registration-processor/core-processor/registration-processor-bio-dedupe-stage/src/test/java/io/mosip/registration/processor/biodedupe/stage/BioDedupeProcessorTest.java (hunk 4 above)
Comment (Khuddusshariff0022) [reply registration-processor/core-processor/registration-processor-bio-dedupe-stage/src/test/java/io/mosip/registration/processor/biodedupe/stage/BioDedupeProcessorTest.java:818]: No. If rid is in PROCESSING state it will be marked to go to Manual Adjudication stage before checking  biometricsUpdateSingleMatchAutoReject.

Comment (Khuddusshariff0022) [reply registration-processor/core-processor/registration-processor-bio-dedupe-stage/src/test/java/io/mosip/registration/processor/biodedupe/stage/BioDedupeProcessorTest.java:820]: If rid is in PROCESSING state then it will marked as send to manual adjudication stage before checking for biometricsUpdateSingleMatchAutoReject.

**Review by ashok-ksharma — APPROVED** [review]

File: registration-processor/core-processor/registration-processor-bio-dedupe-stage/src/test/java/io/mosip/registration/processor/biodedupe/stage/BioDedupeProcessorTest.java (hunk 2 above)
Comment (vishwa-vyom) [reply registration-processor/core-processor/registration-processor-bio-dedupe-stage/src/test/java/io/mosip/registration/processor/biodedupe/stage/BioDedupeProcessorTest.java:719]: If null is not possible why are checking if == null case ?

https://github.com/mosip/registration/blob/11e3ed12384c25d7c28a4e823df255ce86e70fae/registration-processor/core-processor/registration-processor-bio-dedupe-stage/src/main/java/io/mosip/registration/processor/biodedupe/stage/BioDedupeProcessor.java#L420

File: registration-processor/core-processor/registration-processor-bio-dedupe-stage/src/test/java/io/mosip/registration/processor/biodedupe/stage/BioDedupeProcessorTest.java (hunk 4 above)
Comment (vishwa-vyom) [reply registration-processor/core-processor/registration-processor-bio-dedupe-stage/src/test/java/io/mosip/registration/processor/biodedupe/stage/BioDedupeProcessorTest.java:820]: Unit test cases are written not exactly on the code flow, it is to ensure the behavior of the class and it should fail when someone is changing the code later point in time.

In this case, only when biometricsUpdateSingleMatchAutoReject is set to true, there is variation in the behavior of whether to autoreject or send to manual adjudication.

Comment (vishwa-vyom) [reply registration-processor/core-processor/registration-processor-bio-dedupe-stage/src/test/java/io/mosip/registration/processor/biodedupe/stage/BioDedupeProcessorTest.java:820]: Resolving it, for now because this review comment is just on the unit test case, this needs to be discussed later, @Khuddusshariff0022 can we create a task?

File: registration-processor/core-processor/registration-processor-bio-dedupe-stage/src/test/java/io/mosip/registration/processor/biodedupe/stage/BioDedupeProcessorTest.java (hunk 2 above)
Comment (vishwa-vyom) [reply registration-processor/core-processor/registration-processor-bio-dedupe-stage/src/test/java/io/mosip/registration/processor/biodedupe/stage/BioDedupeProcessorTest.java:719]: Resolving it, for now because this review comment is just on the unit test case, this needs to be discussed later, @Khuddusshariff0022 can we create a task?

**Review by vishwa-vyom — APPROVED** [review]
Since we have a time crunch and we have the open review comments only on the unit test case, approving for now and expecting the comments to be handled as a different JIRA task.

Comments to be discussed:
https://github.com/mosip/registration/pull/1901#discussion_r1702148950
https://github.com/mosip/registration/pull/1901#discussion_r1702159828
